src/libAtomVM/bif.c

```diff
@@ -616,6 +619,102 @@ term bif_erlang_sub_2(Context *ctx, uint32_t fail_label, int live, term arg1, te
     }
 }
 
+static inline void intn_to_term_size(size_t n, size_t *intn_data_size, size_t *rounded_num_len)
+{
+    size_t bytes = n * sizeof(intn_digit_t);
+    size_t rounded = ((bytes + 7) >> 3) << 3;
+    *intn_data_size = rounded / sizeof(term);
+    *rounded_num_len = rounded / sizeof(intn_digit_t);
+}
+
+static term make_bigint(Context *ctx, uint32_t fail_label, uint32_t live,
+    const intn_digit_t bigres[], size_t bigres_len)
+{
+    size_t count = intn_count_digits(bigres, bigres_len);
+
+    if (UNLIKELY(count > INTN_MAX_IN_LEN)) {
+        RAISE_ERROR_BIF(fail_label, OVERFLOW_ATOM);
+    }
+
+    if (count > INTN_INT64_LEN) {
+        size_t intn_data_size;
+        size_t rounded_res_len;
+        intn_to_term_size(count, &intn_data_size, &rounded_res_len);
+
+        if (UNLIKELY(
+                memory_ensure_free_with_roots(ctx, BOXED_INTN_SIZE(intn_data_size), live, ctx->x, MEMORY_CAN_SHRINK)
+                != MEMORY_GC_OK)) {
+            RAISE_ERROR_BIF(fail_label, OUT_OF_MEMORY_ATOM);
+        }
+
+        term bigres_term = term_create_uninitialized_intn(intn_data_size, &ctx->heap);
+        intn_digit_t *dest_buf = (void *) term_intn_data(bigres_term);
+        intn_sign_extend(bigres, count, rounded_res_len, dest_buf);
+
+        return bigres_term;
+    } else {
+        int64_t res64 = intn_2_digits_to_int64(bigres, count);
+#if BOXED_TERMS_REQUIRED_FOR_INT64 > 1
+        return make_maybe_boxed_int64(ctx, fail_label, live, res64);
+#else
+        return make_maybe_boxed_int(ctx, fail_label, live, res64);
+#endif
+    }
+}
+
+static void term_to_big_int(term arg1, intn_digit_t *tmp_buf1, intn_digit_t **b1, size_t *b1_len)
+{
+    if (term_is_boxed_integer(arg1)
+        && (term_boxed_size(arg1) > (INTN_INT64_LEN * sizeof(intn_digit_t)) / sizeof(term))) {
+        *b1 = term_intn_data(arg1);
+        *b1_len = term_intn_size(arg1) * (sizeof(term) / sizeof(intn_digit_t));
+    } else {
+        avm_int64_t i64 = term_maybe_unbox_int64(arg1);
+        int64_to_intn_2(i64, tmp_buf1);
+        *b1 = tmp_buf1;
+        *b1_len = INTN_INT64_LEN;
+    }
+}
+
+static void pair_to_big_int(term arg1, term arg2, intn_digit_t *tmp_buf1, intn_digit_t *tmp_buf2,
```

Comment:
I don't understand why this exists as a function. Nothing seems shared between each pair member.